src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,478 @@
+            t.shape.height.set(height);
+        }
+    }
+
+    /// Sets the ports width.
+    pub fn set_width(&self, width:Angle<Degrees>) {
+        self.data.width.set(width);
+        if let Some(t) = self.data.view.data.borrow().as_ref() {
+            t.shape.width.set(width.value.to_radians());
+        }
+    }
+
+    /// Update the position of the sprite according to the Port specification.
+    /// The position is given along a circle, thus the position and rotation of the sprite
+    /// are tied together, so the Port always point in the right direction.
+    /// Needs to be called whenever the `position` or `inner_radius` is modified.
+    fn update_sprite_position(&self) {
+        let inner_radius = self.data.inner_radius.get();
+        let position     = self.data.position.get();
+
+        let translation_vector = Vector3::new(0.0,inner_radius,0.0);
+        let rotation_vector    = -Vector3::new(0.0,0.0,position.rad().value);
+        let rotation           = nalgebra::Rotation3::new(rotation_vector);
+        let translation        = rotation * translation_vector;
+        self.data.view.display_object.set_position(translation);
+        self.data.view.display_object.set_rotation(rotation_vector);
+    }
+
+    /// Update the shape parameters with values from our `Specification`.
+    /// Should only be used upon initialisation.
+    fn init_shape(&self) {
+        if let Some(t) = self.data.view.data.borrow().as_ref() {
+            let height       = self.data.height.get();
+            let inner_radius = self.data.inner_radius.get();
+            let width        = self.data.width.get();
+            let direction    = self.data.direction.get();
+
+            t.shape.update_parameters(height,inner_radius,width,direction);
+        }
+    }
+}
+
+impl<T:PortShapeViewDefinition> Default for Port<T> {
+    fn default() -> Self {
+        Self::new(Specification::default())
+    }
+}
+
+impl<T:PortShapeViewDefinition> From<Specification> for Port<T> {
+    fn from(spec: Specification) -> Self {
+        Self::new(spec)
+    }
+}
+
+
+
+// ===================
+// === Port Buffer ===
+// ===================
+
+/// Data structure that creates and stores port shapes.
+#[derive(Debug,Default)]
+#[allow(missing_docs)]
+pub struct PortBuffer<T:PortShapeViewDefinition> {
+    ports: RefCell<Vec<Port<T>>>,
+}
+
+/// Helper type that represents a `PortBuffer` for `InputPorts`.
+type InputPortBuffer = PortBuffer<InputPortView>;
+/// Helper type that represents a `PortBuffer` for `OutputPorts`.
+type OutputPortBuffer = PortBuffer<OutputPortView>;
+
+impl<T:PortShapeViewDefinition> PortBuffer<T> {
+    /// create a new port in this buffer and sets it as a child node of the given parent.
+    pub fn create(&self, parent:&Node) -> Port<T> {
+        let port = Port::default();
+        parent.add_child(&port.data.view.display_object);
+        self.ports.borrow_mut().push(port.clone());
```

Comment:
port is `CloneRef`, so you sohuld use it here instead to indicate the intention clearly.